src/meta/src/barrier/mod.rs

```diff
@@ -211,40 +337,59 @@ where
     async fn run(&self, mut shutdown_rx: Receiver<()>) {
         let mut tracker = CreateMviewProgressTracker::default();
         let mut state = BarrierManagerState::create(self.env.meta_store()).await;
-
         if self.enable_recovery {
             // handle init, here we simply trigger a recovery process to achieve the consistency. We
             // may need to avoid this when we have more state persisted in meta store.
-            let new_epoch = state.prev_epoch.next();
-            assert!(new_epoch > state.prev_epoch);
-            state.prev_epoch = new_epoch;
+            let new_epoch = state.in_flight_prev_epoch.next();
+            assert!(new_epoch > state.in_flight_prev_epoch);
+            state.in_flight_prev_epoch = new_epoch;
 
             let (new_epoch, actors_to_track, create_mview_progress) =
-                self.recovery(state.prev_epoch).await;
+                self.recovery(state.in_flight_prev_epoch).await;
             tracker.add(new_epoch, actors_to_track, vec![]);
             for progress in create_mview_progress {
                 tracker.update(progress);
             }
-            state.prev_epoch = new_epoch;
-            state.update(self.env.meta_store()).await.unwrap();
+            state.in_flight_prev_epoch = new_epoch;
+            state
+                .update_inflight_prev_epoch(self.env.meta_store())
+                .await
+                .unwrap();
         }
-
         let mut min_interval = tokio::time::interval(self.interval);
         min_interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+        let mut barrier_timer: Option<HistogramTimer> = None;
+        let (barrier_complete_tx, mut barrier_complete_rx) = tokio::sync::mpsc::unbounded_channel();
+        let mut checkpoint_control = CheckpointControl::new();
         loop {
             tokio::select! {
                 biased;
                 // Shutdown
                 _ = &mut shutdown_rx => {
-                    tracing::info!("Barrier manager is shutting down");
+                    tracing::info!("Barrier manager inject is shutting down");
                     return;
                 }
+                result = barrier_complete_rx.recv() => {
+                    let(in_flight_nums,all_nums) = checkpoint_control.get_barrier_len();
+                    self.metrics.in_flight_barrier_nums.set(in_flight_nums as i64);
+                    self.metrics.all_barrier_nums.set(all_nums as i64);
+
+                    if let Some((prev_epoch,result)) = result{
+                        self.barrier_complete_and_commit(prev_epoch,result,&mut state,&mut tracker,&mut checkpoint_control).await;
+                    }
+                    continue;
+                }
                 // there's barrier scheduled.
-                _ = self.scheduled_barriers.wait_one() => {}
+                _ = self.scheduled_barriers.wait_one(), if checkpoint_control.can_inject_barrier(self.in_flight_barrier_nums) => {}
```

Comment:
Possible performance regression, need to think twice. select! will evaluate all the conditions before actually doing selects. Is it possible that:

* can_inject_barrier = false at first
* no in-flight barrier (barrier_complete_rx will stuck)

At this time, only the shutdown branch and barrier_complete_rx will be activated. Even if `can_inject_barrier` becomes true sometime later, the thread will still block on polling `barrier_complete_rx`.

Reply:
Seems that `checkpoint_control` is a local state and will not be modified by other thread. So while the current thread is blocked, it's not likely to be changed.

Reply:
can_inject_barrier will return false in two cases.
 1. the nums of in-flight > max. So there are in-flight barrier
 2. is_build_actor = true. So there is at least one in-flight barrier
So in-flight barrier returns ==>`can_inject_barrier` becomes true. ==> continue next loop

Reply:
Looks like there won't be any deadlock if all other parts are left untouched.